mullvad-daemon fails to delete the wg-mullvad interface after disconnecting

## Issue report

Operating system: Fedora Linux 36, Kernel 5.19.0-0.rc7.53.fc37.x86_64
App version: 2022.3-beta2

### Issue description

mullvad-daemon fails to delete the wg-mullvad interface after disconnecting due to receiving a netlink error on kernel 5.19.x. This is because of a bug in the [rust netlink library](https://github.com/little-dude/netlink) used here https://github.com/mullvad/mullvadvpn-app/blob/baa4802b2e31b55133b3e8fc4ad5058a8ffd0496/talpid-core/Cargo.toml#L59

It got fixed and merged recently: https://github.com/little-dude/netlink/pull/274.

Podman/netavark had to deal with the same issue https://github.com/containers/netavark/issues/345.

> Thanks for the report, we'll try to update the dependency ASAP.

I built master with https://github.com/mullvad/mullvadvpn-app/pull/3775 applied and gave it a test. It unfortunately still fails to remove the interface after a disconnect on 5.19.0-rc7.

```log
mullvad-daemon[22467]: [mullvad_daemon][DEBUG] Target state Secured => Unsecured
mullvad-daemon[22467]: [talpid_core::dns][INFO] Resetting DNS
mullvad-daemon[22467]: [talpid_core::routing::imp::imp][DEBUG] Clearing routes
mullvad-daemon[22467]: [mullvad_daemon][DEBUG] New tunnel state: Disconnecting(Nothing)
mullvad-daemon[22467]: [talpid_core::tunnel::wireguard::wireguard_kernel::netlink_tunnel][ERROR] Failed to remove WireGuard device: Failed to delete device
mullvad-daemon[22467]: [talpid_core::tunnel::wireguard][ERROR] Error: Failed to stop tunnel
mullvad-daemon[22467]: Caused by: Failed to start wireguard tunnel
mullvad-daemon[22467]: [talpid_core::tunnel_state_machine::connecting_state][DEBUG] Tunnel monitor exited with block reason: None
mullvad-daemon[22467]: [talpid_core::firewall][INFO] Resetting firewall policy
mullvad-daemon[22467]: [talpid_core::firewall::imp][DEBUG] Removing table and chain from netfilter
mullvad-daemon[22467]: [mullvad_daemon][DEBUG] New tunnel state: Disconnected
mullvad-daemon[22467]: [mullvad_daemon::management_interface][DEBUG] get_current_location
```

--- a/talpid-core/Cargo.toml
+++ b/talpid-core/Cargo.toml
@@ -56,12 +56,12 @@ jnix = { version = "0.4", features = ["derive"] }
 [target.'cfg(target_os = "linux")'.dependencies]
 inotify = "0.10"
 resolv-conf = "0.7"
-rtnetlink = "0.8"
-netlink-packet-core = "0.2"
-netlink-packet-utils = "0.4"
-netlink-packet-route = "0.8"
-netlink-proto = "0.7"
-netlink-sys = "0.7"
+rtnetlink = "0.11"
+netlink-packet-core = "0.4.2"
+netlink-packet-utils = "0.5.1"
+netlink-packet-route = "0.13"
+netlink-proto = "0.10"
+netlink-sys = "0.8.3"
 nftnl = { version = "0.6.2", features = ["nftnl-1-1-0"] }
 mnl = { version = "0.2.2", features = ["mnl-1-0-4"] }
 which = { version = "4.0", default-features = false }

--- a/talpid-core/src/routing/linux.rs
+++ b/talpid-core/src/routing/linux.rs
@@ -2,6 +2,7 @@ use crate::routing::{
     imp::{CallbackMessage, RouteManagerCommand},
     NetNode, Node, RequiredRoute, Route,
 };
+use netlink_sys::AsyncSocket;
 use std::{
     collections::{BTreeMap, HashSet},
     io,
@@ -144,7 +145,11 @@ impl RouteManagerImpl {
 
         let mgroup_flags = RTMGRP_IPV4_ROUTE | RTMGRP_IPV6_ROUTE | RTMGRP_LINK | RTMGRP_NOTIFY;
         let addr = SocketAddr::new(0, mgroup_flags);
-        connection.socket_mut().bind(&addr).map_err(Error::Bind)?;
+        connection
+            .socket_mut()
+            .socket_mut()
+            .bind(&addr)
+            .map_err(Error::Bind)?;
 
         tokio::spawn(connection);
 

--- a/talpid-core/src/tunnel/wireguard/wireguard_kernel/mod.rs
+++ b/talpid-core/src/tunnel/wireguard/wireguard_kernel/mod.rs
@@ -309,8 +309,7 @@ impl WireguardConnection {
 }
 
 fn consume_netlink_error<
-    T,
-    I: NetlinkDeserializable<T> + Clone + Eq + std::fmt::Debug,
+    I: NetlinkDeserializable + Clone + Eq + std::fmt::Debug,
     F: Fn(rtnetlink::Error) -> Error,
 >(
     message: NetlinkMessage<I>,

--- a/talpid-core/src/tunnel/wireguard/wireguard_kernel/nl_message.rs
+++ b/talpid-core/src/tunnel/wireguard/wireguard_kernel/nl_message.rs
@@ -36,7 +36,7 @@ impl NetlinkControlMessage {
     }
 }
 
-impl NetlinkSerializable<NetlinkControlMessage> for NetlinkControlMessage {
+impl NetlinkSerializable for NetlinkControlMessage {
     fn message_type(&self) -> u16 {
         libc::GENL_ID_CTRL as u16
     }
@@ -57,7 +57,7 @@ impl From<NetlinkControlMessage> for NetlinkPayload<NetlinkControlMessage> {
     }
 }
 
-impl NetlinkDeserializable<NetlinkControlMessage> for NetlinkControlMessage {
+impl NetlinkDeserializable for NetlinkControlMessage {
     type Error = DecodeError;
     fn deserialize(
         _header: &NetlinkHeader,

--- a/talpid-core/src/tunnel/wireguard/wireguard_kernel/wg_message.rs
+++ b/talpid-core/src/tunnel/wireguard/wireguard_kernel/wg_message.rs
@@ -146,7 +146,7 @@ impl DeviceMessage {
     }
 }
 
-impl NetlinkSerializable<DeviceMessage> for DeviceMessage {
+impl NetlinkSerializable for DeviceMessage {
     fn message_type(&self) -> u16 {
         self.message_type
     }
@@ -171,7 +171,7 @@ impl From<DeviceMessage> for NetlinkPayload<DeviceMessage> {
     }
 }
 
-impl NetlinkDeserializable<DeviceMessage> for DeviceMessage {
+impl NetlinkDeserializable for DeviceMessage {
     type Error = Error;
     fn deserialize(header: &NetlinkHeader, payload: &[u8]) -> Result<DeviceMessage, Self::Error> {
         let command = Self::read_genlmsghdr(payload)?;

